File: Graph/js/closestCarrot.js
const closestCarrot=(grid, startRow, startCol)=>{
    const queue = [[startRow, startCol, 0]]
    const visited = new Set([startRow + ',' + startCol])

    while(queue.length > 0){
        const [row, col, distance] = queue.shift()

        if(grid[row][col] === 'C') return distance

        const deltas = [[1,0], [-1,0], [0,1], [0,-1]]

        for(let delta of deltas){
            const [deltaRow, deltaCol] = delta
            const neighborRow = row + deltaRow
            const neighborCol = col + deltaCol

            const neighborPos = neighborRow + ',' + neighborCol

            const rowInbounds = 0 <= neighborRow && neighborRow < grid.length
            const colInbounds = 0 <= neighborCol && neighborCol < grid[0].length

            if(rowInbounds && colInbounds && !visited.has(neighborPos) && grid[neighborRow][neighborCol] !=='X'){
                visited.add(neighborPos)
                queue.push([neighborCol, neighborCol, distance+1])
            }
        }
    }
    return -1
}

const grid = [
  ['O', 'O', 'O', 'O', 'O'],
  ['O', 'X', 'O', 'O', 'O'],
  ['O', 'X', 'X', 'O', 'O'],
  ['O', 'X', 'C', 'O', 'O'],
  ['O', 'X', 'X', 'O', 'O'],
  ['C', 'O', 'O', 'O', 'O'],
];

closestCarrot(grid, 1, 2); // -> 4

File: Graph/go/closest_carrot.go
package graph

type carrotState struct {
	row, col int
	distance int
}

// ClosestCarrot returns the fewest steps from start to any cell containing 'C'.
// Cells marked 'X' are treated as walls.
func ClosestCarrot(grid [][]string, startRow, startCol int) int {
	if len(grid) == 0 {
		return -1
	}
	startPos := positionKey(startRow, startCol)
	queue := []carrotState{{row: startRow, col: startCol, distance: 0}}
	visited := map[string]struct{}{startPos: {}}
	deltas := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

	for len(queue) > 0 {
		state := queue[0]
		queue = queue[1:]

		if grid[state.row][state.col] == "C" {
			return state.distance
		}

		for _, delta := range deltas {
			nr := state.row + delta[0]
			nc := state.col + delta[1]
			if !inBounds(grid, nr, nc) {
				continue
			}
			if grid[nr][nc] == "X" {
				continue
			}
			pos := positionKey(nr, nc)
			if _, seen := visited[pos]; seen {
				continue
			}
			visited[pos] = struct{}{}
			queue = append(queue, carrotState{row: nr, col: nc, distance: state.distance + 1})
		}
	}

	return -1
}
